README.md
The aim of this project is to provide a very simple web form for users
to be able to change their SSH key stored in LDAP or Active Directory
(Samba 4 AD). It’s built with [Bottle](http://bottlepy.org), a WSGI
micro web-framework for Python.

Installation
============

Manually
--------

Clone this repository and install dependencies:

    git clone git@github.com:jceloria/adkey.git
    cd adkey
    pip install -r requirements.txt

Read the next sections to learn how to run it.

Requirements
------------

-   Python 3.x

-   [bottle](https://pypi.python.org/pypi/bottle/)

-   [ldap3](https://pypi.python.org/pypi/ldap3) 2.x

Configuration
=============

Configuration is read from the file
[settings.ini](settings.ini.example). You may change location of the
settings file using the environment variable `CONF_FILE`.

Run it
======

There are multiple ways how to run it:

-   with the built-in default WSGI server based on
    [wsgiref](https://docs.python.org/3/library/wsgiref.html#module-wsgiref.simple_server),

-   under a
    [WSGI](https://en.wikipedia.org/wiki/Web_Server_Gateway_Interface)
    server like [uWSGI](https://uwsgi-docs.readthedocs.org),
    [Waitress](https://docs.pylonsproject.org/projects/waitress),
    [Gunicorn](http://gunicorn.org), … (recommended)

-   as a [CGI](https://en.wikipedia.org/wiki/Common_Gateway_Interface)
    script.

Run with the built-in server
----------------------------

Simply execute the `adkey.py`:

    python3 adkey.py

Then you can access the app on <http://localhost:8080>. The port and
host may be changed in [settings.ini](settings.ini.example).

Run with Waitress
-----------------

    cd adkey
    waitress-serve --listen=*:8080 app:application

Run with uWSGI and nginx
------------------------

If you have many micro-apps like this, it’s IMO kinda overkill to run
each in a separate uWSGI process, isn’t it? It’s not so well known, but
uWSGI allows to “mount” multiple application in a single uWSGI process
and with a single socket.

**Sample uWSGI configuration:.**

    [uwsgi]
    plugins = python3
    socket = /run/uwsgi/main.sock
    chdir = /var/www/scripts
    logger = file:/var/log/uwsgi/main.log
    processes = 1
    threads = 2
    # map URI paths to applications
    mount = /admin/adkey=adkey/adkey.py
    #mount = /admin/change-world=change-world/adkey.py
    manage-script-name = true

**Sample nginx configuration as a reverse proxy in front of uWSGI:.**

    server {
        listen 443 ssl;
        server_name example.org;

        ssl_certificate     /etc/ssl/nginx/nginx.crt;
        ssl_certificate_key /etc/ssl/nginx/nginx.key;

        # uWSGI scripts
        location /admin/ {
            uwsgi_pass  unix:/run/uwsgi/main.sock;
            include     uwsgi_params;
        }
    }

Screenshot
==========

![](doc/screenshot.png)

License
=======

This project is licensed under [MIT
License](http://opensource.org/licenses/MIT/). For the full text of the
license, see the [LICENSE](LICENSE) file.


ChangeLog
CHANGES
=======

* Fix 'Unknown distribution option'
* updated changelog

v2.2.3
------

* updated changelog
* security update

v2.2.2
------

* updated changelog
* security update

v2.2.0
------

* remove extra packages

v2.1.4
------

* let pbr handle changelog
* let pbr handle changelog
* renamed to adkey
* added gunicorn
* use gunicorn with https, move to alpine
* pbr requirement
* add -y

v2.1.3
------

* added pbr to requirements
* add version to failed health check

v2.1.2
------

* updated CHANGELOG
* use pbr for versioning
* added healthcheck
* renamed
* use alpine image instead
* listen on all by default for docker
* update screenshot
* convert to markdown
* expose tcp port
* added trailing slashes
* Added support for Docker
* use bootstrap, update css
* paste ssh key instead of uploading
* use bootstrap, update css
* update ignored files
* update requirements
* extract public key from encrypted private key
* updated docs
* added venv
* use altSecurityIdentities for ssh public key instead of password
* Support any browser error message
* Fix Misspelling in Log Message
* Handle Generic LDAP Errors
* Handle LDAP Socket Connection Errors
* Readme: Add instructions how to install and run on Alpine

v2.0.0
------

* Release v2.0.0
* Readme: Add section Run with Waitress
* Readme: Add Waitress and remove mod\_wsgi from WSGI server tips
* Remove file VERSION
* Allow to enable bottle debug mode using env. variable DEBUG
* Fix code-style
* Add variable VERSION and log app starting
* Move variable BASE\_DIR up
* Use logger module instead of print()
* Rename repository to ldap-passwd-webui
* Fix typo StartSSL -> StartTLS
* Update for ldap3 2.x

v1.0.0
------

* Release v1.0.0
* Add Changelog
* Update Readme, parametrize more strings, remove redundant stuff
* Update year in License file
* Fix version constraint for ldap3 to reject incompatible >= 2.0
* Add example for AD config into settings.ini.example
* Add configuration key "use\_ssl" for LDAP server
* Rename settings.ini to settings.ini.example
* Add configparser to requirements for Legacy Python
* Readme: add warning about AD and unencrypted connection
* Fix form encoding, use getunicode() instead of get()
* Set connection timeout to 5 seconds
* Show "Username or password is incorrect!" when user is not found
* Update year in license
* Add support for ActiveDirectory (Samba 4 AD)
* Add config option search\_filter
* Update readme
* Add robots meta with noindex, nofollow to the page head
* Update readme
* Allow to specify settings.ini location using env. variable
* Allow to configure built-in server via settings.ini
* Improve logging
* First working version
* Initial commit


adkey.py
#!/usr/bin/env python3

import bottle
from bottle import get, post, static_file, request, route, template
from bottle import SimpleTemplate
from configparser import ConfigParser
from ldap3 import Connection, Server
from ldap3 import SIMPLE, SUBTREE, MODIFY_REPLACE
from ldap3.core.exceptions import LDAPBindError, LDAPConstraintViolationResult, \
    LDAPInvalidCredentialsResult, LDAPUserNameIsMandatoryError, \
    LDAPSocketOpenError, LDAPExceptionError
from time import time
from Crypto.PublicKey import RSA
from pbr.version import VersionInfo
import logging
import os


BASE_DIR = os.path.dirname(__file__)
LOG = logging.getLogger(__name__)
LOG_FORMAT = '%(asctime)s %(levelname)s: %(message)s'

_v = VersionInfo(__name__).semantic_version()
__version__ = _v.release_string()
version_info = _v.version_tuple()


@get('/')
def get_index():
    return index_tpl()


@post('/')
def post_index():
    form = request.forms.getunicode

    def error(msg):
        return index_tpl(username=form('username'), alerts=[('error', msg)])

    try:
        key = RSA.importKey(form('ssh-prikey'), passphrase=form('passphrase'))
        ssh_pubkey = key.publickey().export_key('OpenSSH')
    except ValueError as e:
        LOG.error("Unable to decrypt private key for %s: %s" % (form('username'), e))
        return error(str("Unable to decrypt private key"))

    try:
        change_ssh_pubkey(form('username'), form('password'), ssh_pubkey)
    except Error as e:
        LOG.warning("Unsuccessful attempt to change SSH public key for %s: %s" % (form('username'), e))
        return error(str(e))

    LOG.info("SSH public key successfully changed for: %s" % form('username'))

    return index_tpl(alerts=[('success', "SSH public key has been changed")])


@route('/static/<filename>', name='static')
def serve_static(filename):
    return static_file(filename, root=os.path.join(BASE_DIR, 'static'))


@route('/health')
def healthcheck():
    try:
        with connect_ldap(authentication=SIMPLE, user=CONF['ldap']['user'], password=CONF['ldap']['pass']) as c:
            c.bind()
            bottle.response.status = 200
            bottle.response.content_type = 'text/plain'
            return '{} - {}'.format('OK', __version__)

    except (LDAPSocketOpenError, LDAPBindError, LDAPInvalidCredentialsResult):
        bottle.response.status = 503
        bottle.response.content_type = 'text/plain'
        return '{} - {}'.format('FAILED', __version__)


def index_tpl(**kwargs):
    return template('index', **kwargs)


def connect_ldap(**kwargs):
    server = Server(host=CONF['ldap']['host'],
                    port=CONF['ldap'].getint('port', None),
                    use_ssl=CONF['ldap'].getboolean('use_ssl', False),
                    connect_timeout=5)

    return Connection(server, raise_exceptions=True, **kwargs)


def change_ssh_pubkey(*args):
    try:
        if CONF['ldap'].get('type') == 'ad':
            change_ssh_pubkey_ad(*args)
        else:
            change_ssh_pubkey_ldap(*args)

    except (LDAPBindError, LDAPInvalidCredentialsResult, LDAPUserNameIsMandatoryError):
        raise Error('Username or password is incorrect!')

    except LDAPConstraintViolationResult as e:
        # Extract useful part of the error message (for Samba 4 / AD).
        msg = e.message.split('check_password_restrictions: ')[-1].capitalize()
        raise Error(msg)

    except LDAPSocketOpenError as e:
        LOG.error('{}: {!s}'.format(e.__class__.__name__, e))
        raise Error('Unable to connect to the remote server.')

    except LDAPExceptionError as e:
        LOG.error('{}: {!s}'.format(e.__class__.__name__, e))
        raise Error('Encountered an unexpected error while communicating with the remote server.')


def change_ssh_pubkey_ldap(username, passwd, pubkey):
    with connect_ldap() as c:
        user_dn = find_user_dn(c, username)

    # Note: raises LDAPUserNameIsMandatoryError when user_dn is None.
    with connect_ldap(authentication=SIMPLE, user=user_dn, password=passwd) as c:
        c.bind()
        print("Not implemented yet")


def change_ssh_pubkey_ad(username, passwd, pubkey):
    user = username + '@' + CONF['ldap']['ad_domain']
    root = CONF['ldap']['user'] + '@' + CONF['ldap']['ad_domain']
    pubkey = ' '.join(pubkey.decode().split()[:2] + [str(int(time()))])

    # Bind as the requesting user to fetch user_dn
    with connect_ldap(authentication=SIMPLE, user=user, password=passwd) as c:
        c.bind()
        user_dn = find_user_dn(c, username)

    # Use a privileged account to update the attribute
    with connect_ldap(authentication=SIMPLE, user=root, password=CONF['ldap']['pass']) as c:
        c.bind()
        c.modify(user_dn, {'altSecurityIdentities': [(MODIFY_REPLACE, pubkey)]})


def find_user_dn(conn, uid):
    search_filter = CONF['ldap']['search_filter'].replace('{uid}', uid)
    conn.search(CONF['ldap']['base'], "(%s)" % search_filter, SUBTREE)

    return conn.response[0]['dn'] if conn.response else None


def read_config():
    config = ConfigParser()
    config.read([os.path.join(BASE_DIR, 'settings.ini'), os.getenv('CONF_FILE', '')])

    return config


class Error(Exception):
    pass


if os.environ.get('DEBUG'):
    bottle.debug(True)

# Set up logging.
logging.basicConfig(format=LOG_FORMAT)
LOG.setLevel(logging.INFO)
LOG.info("Starting adkey %s" % __version__)

CONF = read_config()

bottle.TEMPLATE_PATH = [BASE_DIR]

# Set default attributes to pass into templates.
SimpleTemplate.defaults = dict(CONF['html'])
SimpleTemplate.defaults['url'] = bottle.url


# Run bottle internal server when invoked directly (mainly for development).
if __name__ == '__main__':
    bottle.run(**CONF['server'])
# Run bottle in application mode (in production under uWSGI server).
else:
    application = bottle.default_app()


app.py
#!/usr/bin/env python3

import bottle
from bottle import get, post, static_file, request, route, template
from bottle import SimpleTemplate
from configparser import ConfigParser
from ldap3 import Connection, Server
from ldap3 import SIMPLE, SUBTREE, MODIFY_REPLACE
from ldap3.core.exceptions import LDAPBindError, LDAPConstraintViolationResult, \
    LDAPInvalidCredentialsResult, LDAPUserNameIsMandatoryError, \
    LDAPSocketOpenError, LDAPExceptionError
import logging
import os
from time import time
from Crypto.PublicKey import RSA
from pbr.version import VersionInfo


BASE_DIR = os.path.dirname(__file__)
LOG = logging.getLogger(__name__)
LOG_FORMAT = '%(asctime)s %(levelname)s: %(message)s'

_v = VersionInfo('mock').semantic_version()
__version__ = _v.release_string()
version_info = _v.version_tuple()


@get('/')
def get_index():
    return index_tpl()


@post('/')
def post_index():
    form = request.forms.getunicode

    def error(msg):
        return index_tpl(username=form('username'), alerts=[('error', msg)])

    try:
        key = RSA.importKey(form('ssh-prikey'), passphrase=form('passphrase'))
        ssh_pubkey = key.publickey().export_key('OpenSSH')
    except ValueError as e:
        LOG.error("Unable to decrypt private key for %s: %s" % (form('username'), e))
        return error(str("Unable to decrypt private key"))

    try:
        change_ssh_pubkey(form('username'), form('password'), ssh_pubkey)
    except Error as e:
        LOG.warning("Unsuccessful attempt to change SSH public key for %s: %s" % (form('username'), e))
        return error(str(e))

    LOG.info("SSH public key successfully changed for: %s" % form('username'))

    return index_tpl(alerts=[('success', "SSH public key has been changed")])


@route('/static/<filename>', name='static')
def serve_static(filename):
    return static_file(filename, root=os.path.join(BASE_DIR, 'static'))


@route('/health')
def healthcheck():
    try:
        with connect_ldap(authentication=SIMPLE, user=CONF['ldap']['user'], password=CONF['ldap']['pass']) as c:
            c.bind()
            bottle.response.status = 200
            bottle.response.content_type = 'text/plain'
            return '{} - {}'.format('OK', __version__)

    except (LDAPSocketOpenError, LDAPBindError, LDAPInvalidCredentialsResult):
        bottle.response.status = 503
        bottle.response.content_type = 'text/plain'
        return '{} - {}'.format('FAILED', __version__)


def index_tpl(**kwargs):
    return template('index', **kwargs)


def connect_ldap(**kwargs):
    server = Server(host=CONF['ldap']['host'],
                    port=CONF['ldap'].getint('port', None),
                    use_ssl=CONF['ldap'].getboolean('use_ssl', False),
                    connect_timeout=5)

    return Connection(server, raise_exceptions=True, **kwargs)


def change_ssh_pubkey(*args):
    try:
        if CONF['ldap'].get('type') == 'ad':
            change_ssh_pubkey_ad(*args)
        else:
            change_ssh_pubkey_ldap(*args)

    except (LDAPBindError, LDAPInvalidCredentialsResult, LDAPUserNameIsMandatoryError):
        raise Error('Username or password is incorrect!')

    except LDAPConstraintViolationResult as e:
        # Extract useful part of the error message (for Samba 4 / AD).
        msg = e.message.split('check_password_restrictions: ')[-1].capitalize()
        raise Error(msg)

    except LDAPSocketOpenError as e:
        LOG.error('{}: {!s}'.format(e.__class__.__name__, e))
        raise Error('Unable to connect to the remote server.')

    except LDAPExceptionError as e:
        LOG.error('{}: {!s}'.format(e.__class__.__name__, e))
        raise Error('Encountered an unexpected error while communicating with the remote server.')


def change_ssh_pubkey_ldap(username, passwd, pubkey):
    with connect_ldap() as c:
        user_dn = find_user_dn(c, username)

    # Note: raises LDAPUserNameIsMandatoryError when user_dn is None.
    with connect_ldap(authentication=SIMPLE, user=user_dn, password=passwd) as c:
        c.bind()
        print("Not implemented yet")


def change_ssh_pubkey_ad(username, passwd, pubkey):
    user = username + '@' + CONF['ldap']['ad_domain']
    root = CONF['ldap']['user'] + '@' + CONF['ldap']['ad_domain']
    pubkey = ' '.join(pubkey.decode().split()[:2] + [int(time())])

    # Bind as the requesting user to fetch user_dn
    with connect_ldap(authentication=SIMPLE, user=user, password=passwd) as c:
        c.bind()
        user_dn = find_user_dn(c, username)

    # Use a privileged account to update the attribute
    with connect_ldap(authentication=SIMPLE, user=root, password=CONF['ldap']['pass']) as c:
        c.bind()
        c.modify(user_dn, {'altSecurityIdentities': [(MODIFY_REPLACE, pubkey)]})


def find_user_dn(conn, uid):
    search_filter = CONF['ldap']['search_filter'].replace('{uid}', uid)
    conn.search(CONF['ldap']['base'], "(%s)" % search_filter, SUBTREE)

    return conn.response[0]['dn'] if conn.response else None


def read_config():
    config = ConfigParser()
    config.read([os.path.join(BASE_DIR, 'settings.ini'), os.getenv('CONF_FILE', '')])

    return config


class Error(Exception):
    pass


if os.environ.get('DEBUG'):
    bottle.debug(True)

# Set up logging.
logging.basicConfig(format=LOG_FORMAT)
LOG.setLevel(logging.INFO)
LOG.info("Starting ldap-ssh-key-webui %s" % __version__)

CONF = read_config()

bottle.TEMPLATE_PATH = [BASE_DIR]

# Set default attributes to pass into templates.
SimpleTemplate.defaults = dict(CONF['html'])
SimpleTemplate.defaults['url'] = bottle.url


# Run bottle internal server when invoked directly (mainly for development).
if __name__ == '__main__':
    bottle.run(**CONF['server'])
# Run bottle in application mode (in production under uWSGI server).
else:
    application = bottle.default_app()
